device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceIO.java

```diff
@@ -365,26 +357,28 @@ public long getReceivePeriodInMilliseconds()
      */
     public void setReceivePeriodInMilliseconds(long newIntervalInMilliseconds) throws IOException
     {
-        /* Codes_SRS_DEVICE_IO_21_030: [If the the provided interval is zero or negative, the setReceivePeriodInMilliseconds shall throw IllegalArgumentException.] */
         if(newIntervalInMilliseconds <= 0L)
         {
             throw new IllegalArgumentException("receive interval can not be zero or negative");
         }
 
-        /* Codes_SRS_DEVICE_IO_21_027: [The setReceivePeriodInMilliseconds shall store the new receive period in milliseconds.] */
         this.receivePeriodInMilliseconds = newIntervalInMilliseconds;
 
-        /* Codes_SRS_DEVICE_IO_21_028: [If the task scheduler already exists, the setReceivePeriodInMilliseconds shall change the `scheduleAtFixedRate` for the receiveTask to the new value.] */
         if (this.receiveTaskScheduler != null)
         {
-            /* Codes_SRS_DEVICE_IO_21_029: [If the `receiveTask` is null, the setReceivePeriodInMilliseconds shall throw IOException.] */
             if (this.receiveTask == null)
             {
                 throw new IOException("transport receive task not set");
             }
 
-            this.receiveTaskScheduler.scheduleAtFixedRate(this.receiveTask, 0,
-                    this.receivePeriodInMilliseconds, TimeUnit.MILLISECONDS);
+            // close the old scheduler and start a new one with the new receive period
+            this.receiveTaskScheduler.shutdown();
```

Comment:
question - when was this old scheduler created? can you set the receive period after client creation or is it set on the transport settings passed in during client creation?

Reply:
These schedulers are created when the connection is first established, and they turn off every time the SDK goes into a reconnection state (since there is no point in trying to send/receive without a connection)

Reply:
It can be set at any time in the DeviceClient lifespan